Hooks/ilhooks.md

```diff
@@ -1,23 +1,369 @@
+to the ldc.r4 instruction.
+
+```csharp
+if (cursor.TryGotoNext
+(
+    i => i.MatchLdcR4(0.7f),
+    i => i.MatchNewobj<WaitForSeconds>()
+))
+{
+    ...
+}
+```
+
+In `GotoNext` and `TryGotoNext`, we define a sequence of instruction predicates (i.e. functions that take in
+`Instruction`s and return `bool`s). All the predicates must match sequential instructions in order. Most instructions
+have a match predicate defined for you to use. In the example above, we can see we are trying to find an `ldc.r4`
+with an operand of 0.7, and a `newobj` creating a `WaitForSeconds` (in this example, there's only one place where
+`ldc.r4 0.7` happens, so you technically don't need the `newobj` check but it helps with readability).
+
+The `Match*` predicates have several variants, depending on how much you care about the operand:
+* If you want to match any `ldc.r4` and need to know the value, you can use `MatchLdcR4(out float value)`
+* If you want to match any `ldc.r4` and don't care about what value, you can use `Match(OpCodes.Ldc_R4)`
+
+These variants apply to other instructions as well. By default, `TryGotoNext` and `GotoNext` place the cursor before
+the matched instructions, so the cursor would be at IL_0057 as shown below. You can change this by setting the first
+argument to `MoveType.After` if needed, but in many cases it will be possible and more convenient to simply match a 
+different set of instructions.
+
+```text
+// <>2__current = new WaitForSeconds(0.7f);
+IL_0056: ldarg.0
+>>>cursor is here<<< IL_0057: ldc.r4 0.7
+IL_005c: newobj instance void [UnityEngine.CoreModule]UnityEngine.WaitForSeconds::.ctor(float32)
+IL_0061: stfld object DropPlatform/'<Flip>d__16'::'<>2__current'
+```
+
+Depending on your preferred mode of failure, you can use `GotoNext` instead of `TryGotoNext`. In this case, you avoid
+an if statement, but if you fail to match IL (i.e. your predicates are incorrect), you'll get an exception. In most
+cases, you're doing a single modification in a known place matching a known pattern, so this may be preferred over
+failing silently. If you need to match several similar chunks of code, you can also use `TryGotoNext` in a while loop
+to make repeated modifications.
+
+### Manipulate Code
+
+Once your cursor is positioned in the right place, you can begin to make modifications to the IL code. This is a good
+time for a reminder that the stack starts empty and needs to end empty - in other words, you need to keep the stack
+balanced. Any time you add a value, you should remove a value, any time you remove a value you should add a value.
+With that in mind, let's look at some common ways to manipulate IL.
+
+`cursor.Next` refers to the instruction the cursor is pointing to. In some cases, it's possible to do simple
+manipulations just by replacing the operand. In the drop platform example, we could simply do 
+`cursor.Next.Operand = NewValue;` inside the if block.
```

Comment:
the new value needs to be an appropriate data type for whatever opcode is there. So if you're doing a ldfld, then yes. If you're doing a ldc.i4, it can just be an int

Reply:
Ah that makes more sense.